handlers/handlers.go

```diff
@@ -0,0 +1,178 @@
+package handlers
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"net/http"
+	"strconv"
+
+	"github.com/codingpop/simplerest/db"
+	"github.com/go-chi/chi"
+)
+
+type store interface {
+	GetPosts(context.Context) ([]db.Post, error)
+	GetPost(context.Context, int) (db.Post, error)
+	CreatePost(context.Context, *db.Post) error
+	UpdatePost(context.Context, int, db.Post) error
+	DeletePost(context.Context, int) error
+}
+
+// Handlers hold all the route handlers
+type Handlers struct {
+	db store
+}
+
+// New creates a new instance of Handlers
+func New(s store) *Handlers {
+	return &Handlers{
+		db: s,
+	}
+}
+
+// GetPosts fetches all posts
+func (h *Handlers) GetPosts(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+
+	posts, err := h.db.GetPosts(r.Context())
+	if err != nil {
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+		return
+	}
+
+	if err := json.NewEncoder(w).Encode(map[string][]db.Post{"posts": posts}); err != nil {
+		http.Error(w, "failed to write response", http.StatusInternalServerError)
+	}
+}
+
+// GetPost fetches a particular post
+func (h *Handlers) GetPost(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+
+	id, err := strconv.Atoi(chi.URLParam(r, "id"))
+	if err != nil {
+		http.Error(w, "invalid id", http.StatusBadRequest)
+		return
+	}
+
+	post, err := h.db.GetPost(r.Context(), id)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			http.Error(w, "post not found", http.StatusNotFound)
+			return
+		}
+
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+		return
+	}
+
+	resp := map[string]db.Post{"post": post}
+
+	if err := json.NewEncoder(w).Encode(resp); err != nil {
+		http.Error(w, "failed to write response", http.StatusInternalServerError)
+	}
+}
+
+// UpdatePost updates a particular post
+func (h *Handlers) UpdatePost(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+
+	var payload struct {
+		Title string `json:"title"`
+		Body  string `json:"body"`
+	}
+
+	id, err := strconv.Atoi(chi.URLParam(r, "id"))
+	if err != nil {
+		http.Error(w, "invalid id", http.StatusBadRequest)
+		return
+	}
+
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&payload); err != nil {
+		http.Error(w, "invalid request body", http.StatusBadRequest)
+		return
+	}
+
+	p := db.Post{
+		ID:    id,
+		Title: payload.Title,
+		Body:  payload.Body,
+	}
+
+	if err := h.db.UpdatePost(r.Context(), id, p); err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			http.Error(w, "post not found", http.StatusNotFound)
+			return
+		}
+
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+		return
+	}
+
+	if err != nil {
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+		return
+	}
```

Comment:
I think this error check is unnecessary?

Reply:
You're right!